pkg/endpointmanager/host.go

```diff
@@ -61,17 +63,37 @@ func (mgr *endpointManager) OnUpdateNode(oldNode, newNode *slim_corev1.Node,
 	oldNodeLabels := oldNode.GetLabels()
 	newNodeLabels := newNode.GetLabels()
 
-	nodeEP := mgr.GetHostEndpoint()
-	if nodeEP == nil {
-		log.Error("Host endpoint not found")
-		return nil
-	}
+	oldNodeAnnotations := oldNode.GetAnnotations()
+	newNodeAnnotations := newNode.GetAnnotations()
 
-	node.SetLabels(newNodeLabels)
+	labelsEqual := comparator.MapStringEquals(oldNodeLabels, newNodeLabels)
+	annotationsEqual := comparator.MapStringEquals(oldNodeAnnotations, newNodeAnnotations)
+
+	// if labels have changed we need to recompute security ID.
+	if !labelsEqual {
+		nodeEP := mgr.GetHostEndpoint()
+		if nodeEP == nil {
+			log.Error("Host endpoint not found")
+			return nil
+		}
+		err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
+		if err != nil {
+			return err
+		}
+	}
 
-	err := nodeEP.UpdateLabelsFrom(oldNodeLabels, newNodeLabels, labels.LabelSourceK8s)
-	if err != nil {
-		return err
+	// Perform a SetMultiAttributes if both labels and annotations has changed,
+	// since these Set operations will wake up Observers on LocalNodeStore its
+	// better to perform one update if multiple attributes change.
+	if !labelsEqual && !annotationsEqual {
+		node.SetMultiAttributes(func(n *node.LocalNode) {
+			n.Labels = newNodeLabels
+			n.Annotations = newNodeAnnotations
+		})
+	} else if !labelsEqual {
+		node.SetLabels(newNodeLabels)
+	} else if !annotationsEqual {
+		node.SetAnnotations(newNodeAnnotations)
```

Comment:
Can we pass the `labelsEqual` and `annotationsEqual` flags to `SetMultiAttribues`, and fold the if-else conditions?

Reply:
That would make SetAnnotations pretty "specific", no? Right now the method doesn't care 'why' you want to set multiple attributes, its just an interface to do so. This seems appropriate to me? Since it may be used in a completely different setting for a completely different evaluation of Node fields.

Reply:
Yes, I do think we *need* an opinionated way to set multiple values on LocalNode without causing LocalNodeStore subscribers to wake up multiple times.